vowpalwabbit/parse_example.cc

```diff
@@ -4,48 +4,85 @@
 
 #include <cmath>
 #include <cctype>
+#include <future>
+#include <thread>
+#include <chrono>
 #include "parse_example.h"
 #include "parse_primitives.h"
 #include "hash.h"
 #include "unique_sort.h"
 #include "global_data.h"
 #include "constant.h"
 #include "vw_string_view.h"
+#include "io_to_queue.h"
 #include "future_compat.h"
 
 #include "io/logger.h"
 
 namespace logger = VW::io::logger;
 
-size_t read_features(vw* all, char*& line, size_t& num_chars)
+size_t read_features(vw *all, std::vector<char>& line, size_t&, v_array<example*>& examples)
 {
-  line = nullptr;
-  size_t num_chars_initial = all->example_parser->input->readto(line, '\n');
-  if (num_chars_initial < 1) return num_chars_initial;
-  num_chars = num_chars_initial;
+  std::vector<char> *io_lines_next_item;
+
+  {
+    std::lock_guard<std::mutex> lck((*all).example_parser->parser_mutex);
+    
+    io_lines_next_item = all->example_parser->io_lines.pop();
+
+    if(io_lines_next_item != nullptr) {
+      (*all).example_parser->ready_parsed_examples.push(examples[0]);
```

Comment:
@peterychang Shouldn't we handle multiple examples here?

Reply:
the examples v_array only ever has a single item in it... with one exception. The json parser can end up having more than one. Since this code is for the text parser, its safe to assume that exactly 1 example will exist in the array